src/c/prace_s_pameti/dynamicka_pamet.md

```diff
@@ -1,6 +1,134 @@
+překladači!). Abychom tomu předešli, tak můžeme použít výraz `sizeof`, kterému můžeme předat datový
+typ[^2] a tento výraz se poté vyhodnotí jako velikost daného datového typu:
+```c,editable,mainbody
+#include <stdio.h>
+int main() {
+    printf("Velikost int je: %d\n", sizeof(int));
+    printf("Velikost int* je: %d\n", sizeof(int*));
+    return 0;
+}
+```
+
+[^2]: Případně výraz, v tom případě si `sizeof` vezme jeho datový typ.
+
+Návratový typ `void*` reprezentuje ukazatel na libovolná data. Funkce `malloc` musí fungovat pro
+alokaci libovolného datového typu, proto musí mít návratový typ právě univerzální ukazatel `void*`.
+Při zavolání funkce `malloc` bychom měli tento návratový typ
+[přetypovat](../datove_typy/celociselne_typy.md#explicitní-konverze) na ukazatel na datový typ,
+který alokujeme.
+
+Při zavolání `malloc`u dojde k naalokování `size` bytů na haldě. Adresa prvního bytu této
+naalokované paměti se poté vrátí jako návratová hodnota `malloc`u. Zde je ukázka programu, který
+naalokuje paměť pro jeden `int` ve funkci, adresu naalokované paměti poté vrátí jako návratovou
+hodnotu a naalokovaná paměť je poté přečtena ve funkci `main`:
+```c,editable
+#include <stdlib.h>
+
+int* allocate() {
+    int* memory = (int*) malloc(sizeof(int));
+    *memory = 5;
+    return memory; 
+}
+int main() {
+    int* memory = allocate();
+    printf("%d\n", *memory);
+    return 0;
+}
+```
+
+### Iniciální hodnota paměti
+Stejně jako u [lokálních proměnných](../promenne/promenne.md#vždy-inicializujte-proměnné) platí, že
+hodnota naalokované paměti je nedefinovaná. Než se tedy hodnotu dané paměti pokusíte přečíst, musíte
+jí nainicializovat zápisem nějaké hodnoty! Jinak bude program obsahovat nedefinované chování 💣.
+
+Pokud byste chtěli, aby naalokovaná paměť byla rovnou při alokaci vynulována (všechny byty
+nastavené na hodnotu `0`), můžete místo funkce `malloc` použít funkci
+[`calloc`](https://devdocs.io/c/memory/calloc)[^3].
+
+[^3]: Pozor však na to, že tato funkce má jiné parametry než `malloc`. Očekává počet hodnot, které
+se mají naalokovat, a velikost každé hodnoty.
 
 ## Uvolnění paměti
+S velkou mocí přichází i velká [zodpovědnost](https://citaty.net/citaty/1957976-stan-lee-s-velkou-moci-prichazi-velka-odpovednost/),
+takže při použití dynamické paměti sice máme více možností než při použití zásobníku, ale zároveň
+MUSÍME tuto paměť korektně uvolňovat (což se u automatické paměti provádělo automaticky). Pokud
+bychom totiž paměť neustále pouze alokovali a neuvolňovali, tak by nám
+[brzy došla](../../caste_chyby/pametove_chyby.md#memory-leak).
+
+Abychom paměť naalokovanou pomocí funkcí `malloc` či `calloc` uvolnili, tak musíme použít funkci
+`free`:
+```c,editable
+#include <stdlib.h>
+
+int main() {
+    int* p = (int*) malloc(sizeof(int)); // alokace paměti
+    *p = 0;                              // použití paměti
+    free(p);                             // uvolnění paměti
```

Comment:
Jeste by se mohlo zminit, ze po `free` by se pointer mel nastavit na `NULL`. Je to sice spise stabni kulture, ale pripada mi to normalni.

Reply:
Já tuhle poučku nemám rád a neřídím se jí :) Pokud to tam napíšeme, tak bych to tam dal jako sekci s trade-offy, abychom jim nenutili jenom jeden názor:
Nastavení na `NULL`:
+ zmenšení šance double free
+ poznáte okamžitě, že ukazatel neukazuje na validní paměť
--- může skrýt double free chybu (popř. use-after-free může nahradit za null dereferenci)

Bez nastavení na `NULL`:
- z hodnoty ukazatele nelze hned poznat, že je paměť uvolněná
+++ při použití ASANu lze odhalit paměťovou chybu (double free)

Paměťové chyby by se neměly skrývat, ale opravit :)

Reply:
OK, nechme tak, kdyz je to dobre pro ASAN, ktery doporucujeme.